### PR TITLE
added test files for array.py

### PR DESCRIPTION
**Pull request by brdgcn**
# changes:
Added test files for 6 different Leetcode problems, each problem are in their own class and method, for future reference we will adding more methods per Leetcode solution for each problem.

please check structure, format, naming conventions, and everything else, please be very picky so I can adopt good programming practices.

### CODE REVIEW

File: data_structure/tests/test_array.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
+import unittest
+from data_structure.array import *
+
+
+class TestPlusOne(unittest.TestCase):
+    def test_plusOne_one(self):
+        output = PlusOne().plusOne([1, 2, 9])
+        ans = [1, 3, 0]
+        self.assertEqual(ans, output)
```
Comment (cheungaryk) [inline data_structure/tests/test_array.py:9]: you want to have something like:
```
self.assertEqual(PlusOne.plusOne([1, 2, 9], [1, 3, 0])
```

so it's easier to read between the 2. Same goes to all the other test cases

File: data_structure/tests/test_array.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    def test_addBinary_two(self):
+        output = AddBinary().addBinary('1010', '1011')
+        ans = "10101"
+        self.assertEqual(ans, output)
+
+
+class TestShortestToChar(unittest.TestCase):
+    def test_ShortestToChar(self):
+        output = ShortestToChar().shortestToChar("loveleetcode", 'e')
```
Comment (cheungaryk) [inline data_structure/tests/test_array.py]: it's not clear what this function do to a reviewer. Typically, your function should explain what it does, but if what it does is too complex to explain in the function name, then you should do inline comment briefly explaining what you expect it to do. Same goes for all the unit test cases below this one (maybe the ones above too, probably).

if you are not entirely sure how to do it, the `given, when, then` structure is a good start. Read https://martinfowler.com/bliki/GivenWhenThen.html for inspirations

File: data_structure/array.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
+class PlusOne(object):
+    # instance method
+    def plusOne(self, digits):
+        carry = 1
+        len_digits = len(digits)
```
Comment (cheungaryk) [inline data_structure/array.py]: len_digits is used only once so it's probably not needed. Just do `for i in reversed(range(len(digits)))`

File: data_structure/array.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+
+class AddBinary(object):
+    def addBinary(self, a, b):
+        # shortcut solution: return bin(int(a, 2) + int(b, 2))[2:]
+        string, carry = '', 0
+        d = len(b) - len(a)
+        a = '0' * d + a
+        b = '0' * -d + b
+        for i, j in zip(a[::-1], b[::-1]):  # zip
```
Comment (cheungaryk) [inline data_structure/array.py]: the word **zip** as a comment is meaningless

File: data_structure/array.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            carry = total // 2
+            print(total, carry, string)
+        return '1' + string if carry else string
+
+
+class ShortestToChar(object):
+    def shortestToChar(self, string, char):
+        charGPS = []
+        res = []
+        len_string = len(string)
```
Comment (cheungaryk) [inline data_structure/array.py]: len_string is not really needed. Just use `len(string)`

File: data_structure/array.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        return 0 if res == len(nums) + 1 else res
+
+
+class MaxChunksToSorted(object):
+    def maxChunksToSorted(self, arr):
+        stack = []
+        for num in arr:
+            if stack and num < stack[-1]:
+                head = stack.pop()
+                while stack and num < stack[-1]: stack.pop()
```
Comment (cheungaryk) [inline data_structure/array.py]: it's more readable to have `while stack... : stack.pop()` in two lines here

```suggestion
                while stack and num < stack[-1]: 
                    stack.pop()
```

**Review by cheungaryk — COMMENTED** [review]
you may want to add some inline comments and docstrings so people know what the functions mean, particularly because the functions are a bit complex. Think about the situation where you leave the code alone for a year then have to come back to read these functions again - you may not understand how the functions work. Write comments to help you understand what it does, how it achieves it, and why you do certain things. Include links if necessary.

Some inspirations for you on how to write good docstrings: https://www.programiz.com/python-programming/docstrings

also, consider using [pylint](https://www.techbeamers.com/pylint-tool/) to analyze your code. There are formatting tools like [black](https://www.freecodecamp.org/news/auto-format-your-python-code-with-black/) that can auto format your code so you don't have to change 1 line at a time. 

afterwards, try using [github actions](https://docs.github.com/en/actions/learn-github-actions) to automatically test and lint your code

File: data_structure/tests/test_array.py (hunk 1 above)
Comment (brdgcn) [reply data_structure/tests/test_array.py:9]: thank you! although line gets really long when it comes to:
        self.assertEqual([True, True, True, False, True], KidsWithCandies().kidsWithCandies([2, 3, 5, 1, 3], 3))
I may keep the current format for more readability (for me at least)

Comment (cheungaryk) [reply data_structure/tests/test_array.py:9]: if you break that line down into multiple lines like this, then it becomes readable:

```
self.assertEqual(
    KidsWithCandies().kidsWithCandies([2, 3, 5, 1, 3], 3),
    [True, True, True, False, True]
)
```

**Comment by brdgcn** [comment]
Thanks very much for all the comments, I made most of the changes, and decide to keep the 
`self.assertEqual(ans, output)`format for now. 
So now that I added some more changes, committed and push to `feedback` branch directly from PyCharm. 
Typically should I request another pull request from `feedback` to `main`, or we can work with the current one?

**Comment by cheungaryk** [comment]
this is how I would do the test cases, using given-when-then (GWT):

```
class TestMinSubArrayLen(unittest.TestCase):
    """
    Given an array of positive integers and a positive integer s, 
    find the minimal length of a contiguous subarray of which the sum ≥ s. 

    If there isn't one, return 0 instead.
    """
    def testContiguousSubarrayLengthOf2(self):
        # given this array and the positive integer
        arr = [2, 3, 1, 2, 4, 3]
        pos_int = 7

        # when I run minSubArrayLen()
        output = MinSubArrayLen().minSubArrayLen(pos_int, arr)

        # then I get an array length of 2 because the last 2 elements are 4 and 3 and they add up to 7
        expected_output = 2

        self.assertEqual(output, expected_output)

    def testTwoOptimalContiguousSubarrays(self):
        # given this array and the positive integer
        arr = [2, 3, 4, 2, 2, 5]
        pos_int = 7

        # when I run minSubArrayLen()
        output = MinSubArrayLen().minSubArrayLen(pos_int, arr)

        # then I get an array length of 2 because 3 + 4 >= 7
        # it doesn't matter that there are 2 optimal contigious subarrays, [3, 4] and [2, 5]
        # because their lengths are the same and both can be used as the output
        expected_output = 2

        self.assertEqual(output, expected_output)

    def testLotsOfZerosInBetween(self):
        # given this array and the positive integer
        arr = [4, 0, 0, 0, 3]
        pos_int = 7

        # when I run minSubArrayLen()
        output = MinSubArrayLen().minSubArrayLen(pos_int, arr)

        # then I get the output 5 because 
        expected_output = 5

        self.assertEqual(output, expected_output)

    def testContiguousSubarrayLengthOf1(self):
        # given this array and the positive integer
        arr = [2, 3, 6, 2, 7, 3]
        pos_int = 7

        # when I run minSubArrayLen()
        output = MinSubArrayLen().minSubArrayLen(pos_int, arr)

        # then I get the output 1 because the 5th element, 7, is by itself equal to the positive integer, 7
        expected_output = 1

        self.assertEqual(output, expected_output)

    def testEmptyArr(self):
        # given this array and the positive integer
        arr = []
        pos_int = 7

        # when I run minSubArrayLen()
        output = MinSubArrayLen().minSubArrayLen(pos_int, arr)

        # then I get the output 0 because the empty array has no value and cannot add up to 7
        expected_output = 0

        self.assertEqual(output, expected_output)

    def testArrCannotAddUpToPositiveInt(self):
        # given this array and the positive integer
        arr = [1, 4, 1]
        pos_int = 7

        # when I run minSubArrayLen()
        output = MinSubArrayLen().minSubArrayLen(pos_int, arr)

        # then I get the output 0 because the positive integer, 7, is higher than 
        # all the integers combined from the array
        expected_output = 0

        self.assertEqual(output, expected_output)

    # add more on your own
```

here's a difference between arrange-act-assert and GWT: https://medium.com/@davembush/nunit-test-code-structure-80b3957550a5. I personally prefer GWT and have only used GWT in the past.

it's your code, so you can do it however you want with it, and if you can, you should also get more reviewers so you can get more opinions on the coding style. My focus is more on readability and making sure that people can understand how the code works even if they have forgotten all about it. I am not that good at algorithms so I wouldn't be able to comment too much on those parts - basically what I am trying to say is that you should write enough comments on those main functions so I don't have to read and analyze too much on your code to understand the logic.

in the future, you will also want to have PRs with fewer changes at a time, so reviewers can focus more on the specifics. For example, you may have 1 function + 1 test case + minor clean-ups in a single PR. In fact, you can try that if you want with new branches and new PRs - by specifying me as the sole reviewer and asking me to be critical on it, I naturally will have a lot of feedback and am reluctant to approve until it's completely adjusted to my liking. 😉

**Pull request merged**